File: zig/src/nucleus/datalog.zig
// Nucleus Datalog Model — SQL generation + execution for DATALOG_* functions
//
// SQL functions: DATALOG_ASSERT, DATALOG_RETRACT, DATALOG_RULE,
// DATALOG_QUERY, DATALOG_CLEAR, DATALOG_IMPORT_GRAPH.

const std = @import("std");
const NucleusClient = @import("client.zig").NucleusClient;

pub const DatalogModel = struct {
    client: *NucleusClient,

    // ── SQL generators ───────────────────────────────────────────

    /// SELECT DATALOG_ASSERT('fact')
    pub fn assertSql(fact: []const u8, buf: []u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "SELECT DATALOG_ASSERT('{s}')", .{fact}) catch return error.BufferTooShort;
    }

    /// SELECT DATALOG_RETRACT('fact')
    pub fn retractSql(fact: []const u8, buf: []u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "SELECT DATALOG_RETRACT('{s}')", .{fact}) catch return error.BufferTooShort;
    }

    /// SELECT DATALOG_RULE('head', 'body')
    pub fn ruleSql(head: []const u8, body: []const u8, buf: []u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "SELECT DATALOG_RULE('{s}', '{s}')", .{ head, body }) catch return error.BufferTooShort;
    }

    /// SELECT DATALOG_QUERY('query')
    pub fn querySql(query_str: []const u8, buf: []u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "SELECT DATALOG_QUERY('{s}')", .{query_str}) catch return error.BufferTooShort;
    }

    /// SELECT DATALOG_CLEAR('predicate'). The argument is required — the
    /// engine answers DATALOG_CLEAR() with "DATALOG_CLEAR requires 1
    /// argument(s), got 0" (measured live; the Go SDK binds the predicate).
    pub fn clearSql(predicate: []const u8, buf: []u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "SELECT DATALOG_CLEAR('{s}')", .{predicate}) catch return error.BufferTooShort;
    }

    /// SELECT DATALOG_IMPORT_GRAPH('predicate') — same arity, verified the
    /// same way: DATALOG_IMPORT_GRAPH() is rejected with "requires 1
    /// argument(s)".
    pub fn importGraphSql(predicate: []const u8, buf: []u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "SELECT DATALOG_IMPORT_GRAPH('{s}')", .{predicate}) catch return error.BufferTooShort;
    }

    // ── Execution methods ────────────────────────────────────────

    pub fn assertFact(self: DatalogModel, fact: []const u8) !?[]const u8 {
        var buf: [1024]u8 = undefined;
        const sql = try assertSql(fact, &buf);
        return try self.client.executeModel(sql);
    }

    pub fn retract(self: DatalogModel, fact: []const u8) !?[]const u8 {
        var buf: [1024]u8 = undefined;
        const sql = try retractSql(fact, &buf);
        return try self.client.executeModel(sql);
    }

    pub fn rule(self: DatalogModel, head: []const u8, body: []const u8) !?[]const u8 {
        var buf: [2048]u8 = undefined;
        const sql = try ruleSql(head, body, &buf);
        return try self.client.executeModel(sql);
    }

    pub fn datalogQuery(self: DatalogModel, query_str: []const u8) !?[]const u8 {
        var buf: [2048]u8 = undefined;
        const sql = try querySql(query_str, &buf);
        return try self.client.executeModel(sql);
    }

    pub fn clear(self: DatalogModel, predicate: []const u8) !?[]const u8 {
        var buf: [256]u8 = undefined;
        const sql = try clearSql(predicate, &buf);
        return try self.client.executeModel(sql);
    }

    pub fn importGraph(self: DatalogModel, predicate: []const u8) !?[]const u8 {
        var buf: [256]u8 = undefined;
        const sql = try importGraphSql(predicate, &buf);
        return try self.client.executeModel(sql);
    }
};

// ── Tests ─────────────────────────────────────────────────────

test "DATALOG_ASSERT sql" {
    var buf: [256]u8 = undefined;
    const sql = try DatalogModel.assertSql("parent(alice, bob)", &buf);
    try std.testing.expectEqualStrings("SELECT DATALOG_ASSERT('parent(alice, bob)')", sql);
}

test "DATALOG_RETRACT sql" {
    var buf: [256]u8 = undefined;
    const sql = try DatalogModel.retractSql("parent(alice, bob)", &buf);
    try std.testing.expectEqualStrings("SELECT DATALOG_RETRACT('parent(alice, bob)')", sql);
}

test "DATALOG_RULE sql" {
    var buf: [512]u8 = undefined;
    const sql = try DatalogModel.ruleSql("ancestor(X,Y)", "parent(X,Y)", &buf);
    try std.testing.expectEqualStrings("SELECT DATALOG_RULE('ancestor(X,Y)', 'parent(X,Y)')", sql);
}

test "DATALOG_QUERY sql" {
    var buf: [256]u8 = undefined;
    const sql = try DatalogModel.querySql("ancestor(alice, X)?", &buf);
    try std.testing.expectEqualStrings("SELECT DATALOG_QUERY('ancestor(alice, X)?')", sql);
}

test "DATALOG_CLEAR sql" {
    var buf: [256]u8 = undefined;
    const sql = try DatalogModel.clearSql("parent", &buf);
    try std.testing.expectEqualStrings("SELECT DATALOG_CLEAR('parent')", sql);
}

test "DATALOG_IMPORT_GRAPH sql" {
    var buf: [256]u8 = undefined;
    const sql = try DatalogModel.importGraphSql("parent", &buf);
    try std.testing.expectEqualStrings("SELECT DATALOG_IMPORT_GRAPH('parent')", sql);
}

File: julia/src/models/datalog.jl
"""Datalog model — logic programming via DATALOG_* SQL functions."""

struct DatalogModel
    conn::LibPQ.Connection
    features::NucleusFeatures
end

"""DATALOG_ASSERT(fact) → String status, e.g. "ASSERT parent/2"

Declared `::Bool` and called `_bool`, but the engine answers with a status
string, so every call raised `MethodError: Cannot convert an object of type
String`. The function had never worked. Rust and Go already returned the string.
"""
function assert_fact!(m::DatalogModel, fact::String)::String
    require_nucleus(m.features, "Datalog")
    result = LibPQ.execute(m.conn, "SELECT DATALOG_ASSERT(\$1)", [fact])
    v = first(result)[1]
    return ismissing(v) ? "" : String(v)
end

"""DATALOG_RETRACT(fact) → String status (same shape as assert_fact!)."""
function retract!(m::DatalogModel, fact::String)::String
    require_nucleus(m.features, "Datalog")
    result = LibPQ.execute(m.conn, "SELECT DATALOG_RETRACT(\$1)", [fact])
    v = first(result)[1]
    return ismissing(v) ? "" : String(v)
end

"""DATALOG_RULE(head, body) → Bool"""
function rule!(m::DatalogModel, head::String, body::String)::Bool
    require_nucleus(m.features, "Datalog")
    result = LibPQ.execute(m.conn, "SELECT DATALOG_RULE(\$1, \$2)", [head, body])
    return _bool(result)
end

"""DATALOG_QUERY(query) → String (CSV results)"""
function datalog_query(m::DatalogModel, query::String)::String
    require_nucleus(m.features, "Datalog")
    result = LibPQ.execute(m.conn, "SELECT DATALOG_QUERY(\$1)", [query])
    val = first(result)[1]
    return ismissing(val) ? "" : val
end

"""DATALOG_CLEAR() → Bool"""
# Takes the predicate to clear.
#
# Sent ZERO arguments where DATALOG_CLEAR requires one, so every call failed
# with "requires 1 argument(s), got 0" — the function had never worked. This is
# the same defect TypeScript had (recorded 2026-08-13), in a second SDK, and
# both survived because nothing executed the call against a live engine.
function clear!(m::DatalogModel, predicate::String)::String
    require_nucleus(m.features, "Datalog")
    result = LibPQ.execute(m.conn, "SELECT DATALOG_CLEAR(\$1)", [predicate])
    v = first(result)[1]
    return ismissing(v) ? "" : String(v)
end

"""DATALOG_IMPORT_GRAPH(predicate) → String status, e.g. "IMPORTED 30 edges into edge"

Sent ZERO arguments where the engine requires one, and declared `::Int64` where
it answers with a status string — two ways of never having worked, in the same
four lines as `clear!`.
"""
function import_graph!(m::DatalogModel, predicate::String)::String
    require_nucleus(m.features, "Datalog")
    result = LibPQ.execute(m.conn, "SELECT DATALOG_IMPORT_GRAPH(\$1)", [predicate])
    v = first(result)[1]
    return ismissing(v) ? "" : String(v)
end
